model/riscv_vmem_utils.sail

```diff
@@ -0,0 +1,261 @@
+/*=======================================================================================*/
+/*  This Sail RISC-V architecture model, comprising all files and                        */
+/*  directories except where otherwise noted is subject the BSD                          */
+/*  two-clause license in the LICENSE file.                                              */
+/*                                                                                       */
+/*  SPDX-License-Identifier: BSD-2-Clause                                                */
+/*=======================================================================================*/
+
+/* This file implements utility functions for accessing memory that
+ * can be used by instruction definitions.
+ *
+ * These memory accessors performs address translation and return a
+ * Retire_Failure on error.  In order to support RVWMO memory model,
+ * address and when possible data arguments are passed as their source
+ * registers.  The `vmem_read` read accessor uses Read(Data) accesses,
+ * while the `vmem_write_*` variants use Write(Data) accesses.
+ *
+ * All addresses are provided in integer source register arguments.
+ * For write accessors, there are functions for each kind of source
+ * data register type (integer, floating point, vector).
+ *
+ * The external API of this module is:
+ *
+ *   check_misaligned - checks for the alignment of a virtual address given platform settings
+ *
+ *   vmem_read        - reads data of the specified width
+ *   vmem_write       - writes a specified data payload
+ */
+
+/* This is a external option that controls the order in which the model
+ * performs misaligned accesses.
+ */
+function sys_misaligned_order_decreasing() -> bool = config memory.misaligned.order_decreasing
+
+/* This is an external option that, when true, causes all misaligned accesses
+ * to be split into single byte operations.
+ */
+function sys_misaligned_to_byte() -> bool = config memory.misaligned.to_byte
+
+/* This is an external option that returns an integer N, such that
+ * when N is greater than zero, misaligned accesses to physical memory
+ * (as atomic events) are allowed provided the access occurs within a N
+ * byte region. N must be a power of two.
```

Comment:
Doesn't it have to be with a *naturally aligned* N byte region?

Also the power of 2 isn't verified anywhere. In fact we should probably just have the exponent as the setting. It would greatly simplify `access_within()`. Something like this I think?

```
function access_within(addr, bytes, region_bytes_exp) = {
  let mask : bits('width) = zero_extend(ones(region_bytes_exp));
  (addr + bytes) & mask == addr & mask
}
```

Reply:
See the other comment.